### DISCUSSION TITLE
LiteFS Cloud is being sunset next month

### DISCUSSION THREAD

**Discussion by MoSattler** [epicweb-dev/epic-stack#858] [General]
Fly io is sunsetting LiteFS Cloud:
> We’ve decided to wind it down. LiteFS Cloud will continue to be available until October 15, 2024. - [Fly.io blog](https://community.fly.io/t/sunsetting-litefs-cloud/20829)

This might require some updates in the readme, like here:
> LiteFS Cloud is a service offered by Fly.io for managing backup and restore functionality. -  [Epic-Stack Database Docs](https://github.com/epicweb-dev/epic-stack/blob/main/docs/database.md#litefs-cloud-backups)

**Comment by kentcdodds** [comment]
Thanks for opening this. I need to evaluate the options and decide which to go with. Got tied up with the Epic React v2 launch. Hopefully I'll have an answer (and maybe migration guide) for what's next before the sunset date.

I'm definitely open to suggestions, especially from folks who have tried alternatives and liked them.

**Reply by MoSattler** [reply]
@benbjohnson made some recommendations:

> ### Disaster Recovery Without LiteFS Cloud
>
> There are two good options.
>
> #### LiteFS + Litestream
>
> You can use [Litestream](https://litestream.io/) to maintain your own backups to an S3-compatible object storage such as [Tigris](https://www.tigrisdata.com/). That’s what we do internally. If you’re running a static primary, you can point Litestream at your primary node in your cluster.
>
> #### Turso
>
> If you’re interested in running a fully-managed SQLite service, we like [Turso](https://turso.tech/). You can find more information on their [Quick Start](https://docs.turso.tech/quickstart) page.

**Reply by buildtheui** [reply]
Hi everyone, I wanted to share my experience migrating to static leasing with Litestream and Tigris. Although I'm not using the Epic Stack, I was heavily inspired by it to build my backend in Golang using LiteFS and LiteFS Cloud. Since the concepts are similar, I thought it would be helpful to share my progress.

I followed the short migration guide from Fly.io and encountered a few challenges along the way. [I’ve documented the issues here](https://community.fly.io/t/error-with-read-only-replica-configuration-in-litefs-cloud-migration-to-litestream-tigris/22209). However, I’ve hit a roadblock: I can't figure out how to set the `lease.hostname` to the primary machine ID for all nodes (both primary and replicas).

I’m not a Fly.io expert, so if anyone has a better approach or suggestions, this would help everyone.

**Reply by janhesters** [reply]
@buildtheui where you able to solve this issue?

**Comment by buildtheui** [comment]
As another option, @stephen built a drop-in solution for Litefs cloud, more details [here](https://community.fly.io/t/litefs-backup-drop-in-replacement-for-litefs-cloud/22267)

**Reply by stephen** [reply]
👋 [litefs-backup](https://github.com/stephen/litefs-backup) is a stripped down fork of the original LiteFS Cloud implementation, so you're self-hosting your backup solution.

My primary motivation was to continue to have support for LiteFS' dynamic/consul leasing.

**Reply by NGimbal** [reply]
I just deployed litefs-backup and it worked well. The documentation was clear and correct. Thanks @stephen!

**Comment by MoSattler** [comment]
Looks like so far LiteFS Cloud is still available 🤔

**Reply by MoSattler** [reply]
Ah really? which ones?

**Comment by timothymalcham** [comment]
Just checking to see if anyone has used litefs-backup or another solution successfully with the Epic Stack? Did it work as a drop-in replacement, any issues? I've got a couple client projects built with the Epic Stack and would like to implement something automated for backups.

**Reply by MoSattler** [reply]
I have bitten the bullet and migrated to Postgres :(

**Comment by davidesigner** [comment]
Hi everyone! 

Have any of you implemented Turso in the Epic Stack?

I found this example: https://github.com/tursodatabase/epic-stack made by @giovannibenussi, but I'm not sure I understand the migration process:
- In the local environment: I see: `cat prisma/migrations/20230914194400_init/migration.sql | sqlite3 ./data.db`, so if I understand, the migration applied by Prisma when we do `prisma migrate reset` is not applied to the local database?
- In production environment: I see `turso db shell epic-stack-example < prisma/migrations/20230914194400_init/migration.sql`, so we need to manually apply migrations to the main Turso database in production? Maybe we should create a service that automatically applies migrations during the deployment step in Github Actions?

I have other questions too, is LiftFS still recommended when using Turso?

Any help would be appreciated, thanks 🙏

**Reply by giovannibenussi** [reply]
Hey, Giovanni from Turso here 🙋
- You're right, `prisma migrate reset` is not applied to the local database so you'd need to do this manually.
- Same, you need to do this manually in production. 

We need to coordinate with the Prisma team to improve this whole process.

Regarding LiteFS, you don't need it when using Turso. You can create a Turso account and don't worry about servers anymore.
Also, if you're on an environment with access to the file-system (like Fly), you can use [Embedded Replicas](https://docs.turso.tech/sdk/ts/reference#embedded-replicas) to replicate your database in your server to enable microsecond-level read operations.

Let me know if you need help with anything else or if there's anything that we can improve 😄

**Reply by davidesigner** [reply]
Hi Giovanni, thanks for your reply.

Are you planning to improve the sync with Prisma in the near future? If so, when?

I certainly won't be using Turso until it's implemented, because I couldn't just have a smooth deployment with no downtime.

Without it, I'd have to develop a migration management overlay...

I think Turso is definitely a good alternative if it can be easily implemented in the Epic stack.

Have a nice day,
David

**Reply by giovannibenussi** [reply]
Hey, I no longer work at Turso so I'm not sure about timelines. I'd ask in [their Discord](https://discord.gg/turso) to get a better answer!

**Comment by kentcdodds** [comment]
I still need to add a built-in solution for this, but now that we have Tigris support built-into the Epic Stack, we could probably use litestream to handle auto-backups. I think @devneill has done something similar to s3.

**Reply by devneill** [reply]
I'll take a look at this early next week and see if I can help 🧐

**Comment by devneill** [comment]
Hopefully this can help.

I managed to add backups via Github actions and Tigris.

**Github action to trigger the backup daily:**
https://github.com/devneill/getfranki-08-24/blob/main/.github/workflows/daily-backup.yml

**Resource endpoint in my app to handle the request:**
https://github.com/devneill/getfranki-08-24/blob/main/app/routes/api%2B/backup.ts

**The endpoints calls the backDB() function in here:**
https://github.com/devneill/getfranki-08-24/blob/main/app/utils/db.server.ts

**Which calls this script:**
https://github.com/devneill/getfranki-08-24/blob/main/other/backup_db.sh

That script makes a copy of the db and sends it to an s3 bucket, which is in Tigris.

---

I will open a PR to add a better version of this to the Epic Stack once I have more time (after the Epic Web Conference.)

**Reply by kentcdodds** [reply]
This is awesome. Thank you for sharing @devneill!

**Reply by taylrd** [reply]
Thank you for sharing @devneill :)

**Reply by NGimbal** [reply]
Thanks for sharing :) Have you worked out the process for restoring from a backup created this way?

**Reply by devneill** [reply]
Hey @NGimbal

Yes! There is a similar flow to the one above.

Starting from this GH Action:
https://github.com/devneill/getfranki-08-24/blob/main/.github/workflows/emergency-restore.yml

Which hits this resource route:
https://github.com/devneill/getfranki-08-24/blob/main/app/routes/api%2B/restore.ts

**Reply by NGimbal** [reply]
Thanks for the response!

**Reply by janhesters** [reply]
@devneill May I ask what the status of this is? What would you recommend nowadays?

**Comment by NGimbal** [comment]
Just want to say that as someone who's been building on the Epic Stack for ~18 months this problem isn't solved. I'm looking to move to another database option just because the backup and restore situation is too flaky and not well supported. My personal saga is:

- I've been using `lfsb` for about a month which was working well, although I tried to restore to a specific checkpoint and borked my database. Thanks to `lfsb` I had a recent backup and was able to restore the app.
- I'm now going to @devneill's solution because it's a little more straightforward. But the solution above is more of a (good) sketch, and is not as thorough as the rest of Epic Stack (maybe as expected). I'm working through it though.

It seems that Litestream with static leases is the way that Fly is moving and maybe should be the Epic Stack recommendation in the future when someone has the time to put it together and test it. For me personally I think I'm going to move to a managed Postgres solution so that I can let others manage the backup / restore functionality, and because the Postgres ecosystem has a lot of really nice extensions I'd like to use.

Just my two cents in case it's helpful to the discussion.

**Reply by kentcdodds** [reply]
Thanks for putting in your experience here! I think right now the thing I would recommend is @devneill's approach for now.

**Reply by NGimbal** [reply]
Thanks for the reply. I've implemented the Github Actions solution. I think I'm actually going to look further into Litestream since it seems like that might be better supported by Fly going forward. I'm going to hold off on Postgres a bit longer.

**Reply by janhesters** [reply]
@NGimbal did you end up going with Litestream? AFAIK it does NOT yet support read replicas. So did you keep LiteFS as read replicas?

**Reply by NGimbal** [reply]
I went with the method described above, LiteFS with GitHub actions backup and restore. That's worked well for a while.

**Reply by janhesters** [reply]
@NGimbal Thank you for answering 🙏

**Comment by janhesters** [comment]
Did anyone figure out an alternative or how to fix this in the Epic Stack?

**Reply by NGimbal** [reply]
It seems like the Fly.io blessed approach is Litestream although I haven't successfully migrated to Litestream. I'm still using Litefs with backup / restore as described above. It's worked, I backup daily with a GH action and I have restored the database 1-2 times successfully. Fly.io also has a managed Postgres option now. I don't think there's an official Epic stack solution to this problem.

**Reply by janhesters** [reply]
@NGimbal Thank you 🙏 

Litestream seem a lot more complicated than the LiteFS cloud was ...
